src/main/java/teammates/ui/webapi/DeadlineExtensionsWorkerAction.java

```diff
@@ -0,0 +1,156 @@
+package teammates.ui.webapi;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import teammates.common.datatransfer.attributes.CourseAttributes;
+import teammates.common.datatransfer.attributes.DeadlineExtensionAttributes;
+import teammates.common.datatransfer.attributes.FeedbackSessionAttributes;
+import teammates.common.exception.EntityAlreadyExistsException;
+import teammates.common.exception.EntityDoesNotExistException;
+import teammates.common.exception.InvalidParametersException;
+import teammates.common.util.EmailType;
+import teammates.common.util.EmailWrapper;
+import teammates.common.util.Logger;
+import teammates.ui.request.DeadlineExtensionRequest;
+import teammates.ui.request.InvalidHttpRequestBodyException;
+
+/**
+ * Task queue worker action: creates, updates or deletes deadline extensions and sends out appropriate emails.
+ */
+class DeadlineExtensionsWorkerAction extends AdminOnlyAction {
+
+    private static final Logger log = Logger.getLogger();
+
+    @Override
+    public JsonResult execute() throws InvalidHttpRequestBodyException {
+        DeadlineExtensionRequest deadlineExtensionRequest = getAndValidateRequestBody(DeadlineExtensionRequest.class);
+        CourseAttributes course = logic.getCourse(deadlineExtensionRequest.getCourseId());
+        FeedbackSessionAttributes session = logic.getFeedbackSession(deadlineExtensionRequest.getFeedbackSessionName(),
+                deadlineExtensionRequest.getCourseId());
+
+        if (course == null || session == null) {
+            log.severe("Course with id: " + deadlineExtensionRequest.getCourseId() + " or feedback session: "
+                    + deadlineExtensionRequest.getFeedbackSessionName() + " not found when updating deadline extensions");
+            return new JsonResult("Failure");
+        }
+
+        boolean notifyUsers = deadlineExtensionRequest.getNotifyUsers();
+
+        createOrUpdateDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getStudentExtensionsToModify(), false, notifyUsers);
+        createOrUpdateDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getInstructorExtensionsToModify(), true, notifyUsers);
+        revokeDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getStudentExtensionsToRevoke(), false, notifyUsers);
+        revokeDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getInstructorExtensionsToRevoke(), true, notifyUsers);
+
```

Comment:
Also I'm not sure about this, it's probably quite minor, but could we benefit from collating all the emails together and scheduling them in one shot? In terms of having a better spacing between emails

Reply:
I think we can! Actually, originally that was how I designed it but it was really messy. I've simplified it quite a bit since my first attempt, will try again to see if it's possible.

Reply:
Should be resolved, can see if it's ok.